Dropdown question type, ref #340

jancborchardt opened the pull request:
As per #340, starting with the dropdown question type! :)

To do:
- [x] In the Create.vue, in the "Edit" mode it’s fine to list all the answers, but for viewing it should switch into an actual dropdown
- [x] Same for the share link view

skjnldsv commented on src/components/Questions/QuestionMultiple.vue:
```diff
@@ -73,7 +73,11 @@
 
 			<li v-if="(edit && !isLastEmpty) || hasNoAnswer" class="question__item">
-				<div class="question__item__pseudoInput" :class="{'question__item__pseudoInput--unique':isUnique}" />
+				<div class="question__item__pseudoInput"
+					:class="{
+						'question__item__pseudoInput--unique':isUnique,
+						'question__item__pseudoInput--uniqueDropdown':isDropdown
+					}" />
```
I think it's best to have its own component :)

skjnldsv replied:
This one is too complicated because designed for the multiple checkbox/radio logic ;)

jancborchardt replied:
Done, with a separate component that could even be extended in the future to support a `<select>` with the `multiple` attribute. :) But not for now as we don’t even have any styles for that anyway and it’s out of scope anyhow.

jotoeri commented on src/components/Questions/QuestionDropdown.vue:
```diff
@@ -0,0 +1,398 @@
+	display: inline-flex;
+	min-height: 44px;
+
+	.question__label {
+		flex: 1 1 100%;
+		// Overwrite guest page core styles
+		text-align: left !important;
+		// Some rounding issues lead to this strange number, so label and answerInput show up a the same position, working on different browsers.
+		padding: 6.5px 0 0 30px;
+		line-height: 22px;
+		min-height: 34px;
+		height: min-content;
+		position: relative;
+
+		&::before {
+			box-sizing: border-box;
+			// Adjust position manually for proper position to text
+			position: absolute;
+			top: 10px;
+			width: 16px;
+			height: 16px;
+			margin-bottom: 0;
+			margin-left: -30px !important;
+			margin-right: 14px !important;
+		}
+	}
```
Unused copyPaste? ;)

jotoeri commented on src/components/Questions/QuestionDropdown.vue:
```diff
@@ -0,0 +1,398 @@
+	position: relative;
+}
+
+input.question__radio,
+input.question__checkbox {
+	z-index: -1;
+	// make sure browser warnings are properly
+	// displayed at the correct location
+	left: 0px;
+	width: 16px;
+}
```
copyPaste...

jancborchardt commented on src/models/AnswerTypes.js:
```diff
@@ -59,13 +71,15 @@ export default {
-		label: t('forms', 'Checkboxes'),
+	dropdown: {
+		component: QuestionDropdown,
+		icon: 'icon-triangle-s',
```
Yup! Will work on a specific question type icon for that, this was mostly a placeholder. :)

jancborchardt commented on src/components/Questions/AnswerInput.vue:
```diff
@@ -204,6 +212,11 @@ export default {
 		}
 
+		// Do not show pseudo-icon for dropdowns
+		&--dropdown {
```
I think it’s ok to call it --dropdown in this case since that’s what it’s for, and it could have more rules in the future.

jancborchardt commented on src/models/AnswerTypes.js:
```diff
@@ -20,14 +20,16 @@
-import QuestionLong from '../components/Questions/QuestionLong'
-import QuestionShort from '../components/Questions/QuestionShort'
 import QuestionMultiple from '../components/Questions/QuestionMultiple'
+import QuestionDropdown from '../components/Questions/QuestionDropdown'
+import QuestionShort from '../components/Questions/QuestionShort'
+import QuestionLong from '../components/Questions/QuestionLong'
```
Hmm right – I was a bit confused since these things were ordered differently in different places, so I ordered them the way we want to order them in the interface to have symmetry there.

jancborchardt commented:
@skjnldsv how should we move forward here – do we continue to work it into its own component, or use QuestionMultiple?

The editing is basically the same except the pseudo radio/checkboxes (we could do numbering instead just to have something), and the viewing is the `<select>` dropdown. Most of the component is copied so it would still make sense to just use the same component.

Would be good if you can give it a check what makes most sense.

skjnldsv commented:
@jancborchardt Own component is good.
As you said it's best for future use as this is probably gonna give us some more flexibility later on :)

jancborchardt commented:
@skjnldsv cool, sounds good. Then I’d fix the stuff from @jotoeri’s review? Please check too if there’s other things to fix as well.

jancborchardt commented on src/components/Questions/QuestionDropdown.vue:
```diff
@@ -0,0 +1,398 @@
+			</option>
+		</select>
+
+		<ol v-if="edit" class="question__content">
```
Both should use ol, since it is indeed an _ordered_ list. So fixing it here, we can adjust the QuestionMultiple later as well.

jancborchardt commented on src/components/Questions/QuestionDropdown.vue:
```diff
@@ -0,0 +1,398 @@
+		@update:mandatory="onMandatoryChange"
+		@delete="onDelete">
+		<select v-if="!edit"
+			:id="text"
+			:name="text"
```
From the docs https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select
> The above example shows typical <select> usage. It is given an id attribute to enable it to be associated with a <label> for accessibility purposes, as well as a name attribute to represent the name of the associated data point submitted to the server.

jancborchardt commented on src/components/Questions/AnswerInput.vue:
```diff
@@ -54,6 +58,10 @@ export default {
 			type: Boolean,
 			required: true,
 		},
+		isDropdown: {
+			type: Boolean,
+			required: true,
+		},
```
Fixed by passing `:is-dropdown="false"` in QuestionMultiple. :)

jancborchardt commented:
> Ok, so some comments i wrote on code, some general points i have still here:

Fixed them all! :) Is the code fine now @skjnldsv @rullzer @jotoeri?

> Can we create some larger icon? This one feels quite small... (Maybe also with frame, to get that typical dropdown-box?)

I’ll work on this. :)

> Would be good to adjust padding & height on Submit.vue & Create.vue to get the same distance to text, as on other question-types?

I’ll check that out as well.

> Somehow connected to this, on Create.vue, ideally, the drag-handle and as well as the text would not move anymore, if changing to edit-mode. (if possible ;))

I would say it’d be good if the drag-handle is just always shown next to the title instead of in the middle of the question. This will also help later when we implement the ability to drag&drop _answers_ and we need to have the drag handles there too.
But the handle placement issue is separate from this I’d say?

> Maybe would be cool, to show the dropdown normally, if specifically clicking the arrow on Create? So it is somehow possible to see the Dropdown, like users will see... That was at least the goal on all other question-types...?

Good point but not sure if it’s possible.

> Dropdown still has a large scroll-bar, not a thin one as everywhere on NC?

I’ll check. :)

jancborchardt commented:
> Can we create some larger icon? This one feels quite small... (Maybe also with frame, to get that typical dropdown-box?)

Done, added a proper icon for it. :)

> - Would be good to adjust padding & height on Submit.vue & Create.vue to get the same distance to text, as on other question-types?
> - Dropdown still has a large scroll-bar, not a thin one as everywhere on NC?

Adjusted the dropdown styles as good as possible. The options styling is very limited, see https://stackoverflow.com/questions/7208786/how-to-style-the-option-of-an-html-select-element
E.g. it’s unfortunately not possible to give the `<option>` elements any padding as they are handled by the operating system.

> - Somehow connected to this, on Create.vue, ideally, the drag-handle and as well as the text would not move anymore, if changing to edit-mode. (if possible ;))
> - Maybe would be cool, to show the dropdown normally, if specifically clicking the arrow on Create? So it is somehow possible to see the Dropdown, like users will see... That was at least the goal on all other question-types...?

I’d say these 2 are separate from this.

rullzer commented:
Still marked as developing? is that right?

jancborchardt commented:
@rullzer no, missed that – it’s ready to review from my side.

jancborchardt commented:
> I would prefer the use of our Multiselect component but it's perfect for 2.0.0! :)
> @jancborchardt, could you create a dedicated issue for this after this is merged?

Because the call on whether to use the multiselect component or its own component has changed around so much, I’d prefer if you can create that follow-up issue with the relevant info. :)

(And yes, I’d say it makes sense to use the multiselect component, especially as it’s probably easier once we want the ability to change question types and keep the question & answers text intact.)

skjnldsv commented:
> Because the call on whether to use the multiselect component or its own component has changed around so much, I’d prefer if you can create that follow-up issue with the relevant info. :)

Roger roger!

jancborchardt commented:
Rebased and ready to go as well! :)
